lib/galaxy/model/migrations/alembic/versions_gxy/d619fdfa6168_username_column_unique_constraint.py

```diff
@@ -0,0 +1,45 @@
+"""Username column unique constraint
+
+Revision ID: d619fdfa6168
+Revises: eee9229a9765
+Create Date: 2024-07-02 13:13:10.325586
+
+Upgrade note: This upgrade will raise an error if there are duplicate username
+values in the galaxy_user table. In that case, duplicate values should be
+removed. For that, run the username deduplication script availbable in the data
+package.
+"""
+
+from galaxy.model.database_object_names import (
+    build_index_name,
+    build_unique_constraint_name,
+)
+from galaxy.model.migrations.util import (
+    create_index,
+    create_unique_constraint,
+    drop_constraint,
+    drop_index,
+    transaction,
+)
+
+# revision identifiers, used by Alembic.
+revision = "d619fdfa6168"
+down_revision = "eee9229a9765"
+branch_labels = None
+depends_on = None
+
+table_name = "galaxy_user"
+column_name = "username"
+constraint_name = build_unique_constraint_name(table_name, [column_name])
+index_name = build_index_name(table_name, [column_name])
+
+
+def upgrade():
+    with transaction():
+        create_unique_constraint(constraint_name, table_name, [column_name])
```

Comment:
This is going to fail without https://github.com/galaxyproject/galaxy/pull/18492. I would pull that into the migration. Once the constraint is present we'll enver need the script, so I don't see why we have it live as a separate script.

Reply:
My initial thinking was that we could run that script on the current release, which would deduplicate the usernames and, with that, stop the errors. However, the more I thought about it, I became convinced that it shouldn't be merged into a stable release: the script is a feature and belongs in dev. So yes, I'll move it into the same PR. 

That said, I'm not sure we should be combining database schema revisions with data migrations in the same revision script. I think it may be better to have a script and reference it in the db migration module as an "upgrade note" (like I did here) + in the release admin notes.

Reply:
OK. I suppose we can distinguish between (a) a data migration that addresses business logic needs and (b) a data migration that fixes inconsistent data to enable the db schema migration; and combine the latter with the db schema migration into one revision script. We should then mention such cases in the release admin notes to alert admins that some data is being changed (adding a label now).